cpp/ql/src/semmle/code/cpp/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -82,6 +82,21 @@ predicate localAdditionalTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeT
   exprToDefinitionByReferenceStep(nodeFrom.asExpr(), nodeTo.asDefiningArgument())
   or
   exprToPartialDefinitionStep(nodeFrom.asExpr(), nodeTo.asPartialDefinition())
+  or
+  // Reverse taint: if a function model has taint from the qualifier to the
+  // dereferenced return value, also apply reverse taint from the post-update
+  // of the return value back to the qualifier. This allows taint to flow 'in'
+  // through references returned by a modeled function such as `operator[]`.
```

Comment:
> in the long run we may have to write models specifically about aliasing since from a taint model it's impossible to tell whether the data was copied or aliased.

Perhaps what I'm suggesting here is exactly what you proposed originally: specifying flow _from_ `isReturnValueDeref` _to_ the qualifier object or another parameter. Then backwards flow means aliasing.

Reply:
This would also be fixed with explicit reverse taint declared in each taint model where it applies.  In the case of `operator+=` I think that would mean from the returned reference to the LHS (qualifier) but not the RHS.  I know you don't like it, but I'm still not clear why we don't like reverse taint in models.

Reply:
I think I'll try out the explicit reverse taint approach and see how the solutions compare...

Reply:
I've just pushed a commit ("C++: Explicitly model data flow in through reference return values.") that changes strategy as discussed here.  I'd appreciate your eyes on it.

Reply:
It's taken me a long time to accept, but I think your concept of flow _from_ the returned object is what we need. Let's go with that.